### PR TITLE
r/dynamodb_resource_policy

### LINKED ISSUE #36502: Amazon DynamoDB resource-based policies
> Amazon DynamoDB now supports resource-based policies to help you simplify access control for your DynamoDB resources.

* [Announcement](https://aws.amazon.com/about-aws/whats-new/2024/03/amazon-dynamodb-resource-based-policies/)
* [Blog post](https://aws.amazon.com/blogs/database/simplify-cross-account-access-control-with-amazon-dynamodb-using-resource-based-policies/)

Requires [AWS SDK for Go v1.51.4](https://github.com/aws/aws-sdk-go/releases/tag/v1.51.4): https://github.com/hashicorp/terraform-provider-aws/pull/36497.

### Potential Terraform Configuration

```terraform
resource "aws_dynamodb_resource_policy" "example" {
  resource_arn = aws_dynamodb_table.example.arn

  policy = <<EOF
...
EOF
}
```

### PR DESCRIPTION
**Pull request by danquack**
### Description

Adds the ability to set a resource policy on a dynamodb table.

### Relations

Closes #36502

### Output from Acceptance Testing

```console
% TF_ACC=1 make testacc TESTS=TestAccDynamoDBResourcePolicy_  PKG=dynamodb                             
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.21.8 test ./internal/service/dynamodb/... -v -count 1 -parallel 20 -run='TestAccDynamoDBResourcePolicy_'  -timeout 360m
=== RUN   TestAccDynamoDBResourcePolicy_basic
=== PAUSE TestAccDynamoDBResourcePolicy_basic
=== RUN   TestAccDynamoDBResourcePolicy_update
=== PAUSE TestAccDynamoDBResourcePolicy_update
=== RUN   TestAccDynamoDBResourcePolicy_disappears
=== PAUSE TestAccDynamoDBResourcePolicy_disappears
=== CONT  TestAccDynamoDBResourcePolicy_basic
=== CONT  TestAccDynamoDBResourcePolicy_disappears
=== CONT  TestAccDynamoDBResourcePolicy_update
--- PASS: TestAccDynamoDBResourcePolicy_basic (39.19s)
--- PASS: TestAccDynamoDBResourcePolicy_update (54.47s)
--- PASS: TestAccDynamoDBResourcePolicy_disappears (110.16s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   114.778s
```

### CODE REVIEW

**Comment by danquack** [comment]
This was done with skaff, maybe I removed too many comments.

File: internal/service/dynamodb/resource_policy.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"id": framework.IDAttribute(),
+			"policy": schema.StringAttribute{
+				CustomType: fwtypes.IAMPolicyType,
+				Required:   true,
+			},
+			"confirm_remove_self_resource_access": schema.BoolAttribute{
```
Comment (danquack) [inline internal/service/dynamodb/resource_policy.go:74]: This seems like an overwrite flag. I could see the provider removing in, in favor of import logic, but would love to get someone else's opinion

Comment (ewbankkit) [reply internal/service/dynamodb/resource_policy.go:74]: My reading of the [documentation](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_PutResourcePolicy.html#DDB-PutResourcePolicy-request-ConfirmRemoveSelfResourceAccess) is that if you are trying to put a resource policy that doesn't `ALLOW` the caller to call `PutResourcePolicy` in the future then you need to set this flag to make that happen.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccDynamoDBResourcePolicy_' PKG=dynamodb
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.21.8 test ./internal/service/dynamodb/... -v -count 1 -parallel 20  -run=TestAccDynamoDBResourcePolicy_ -timeout 360m
=== RUN   TestAccDynamoDBResourcePolicy_basic
=== PAUSE TestAccDynamoDBResourcePolicy_basic
=== RUN   TestAccDynamoDBResourcePolicy_update
=== PAUSE TestAccDynamoDBResourcePolicy_update
=== RUN   TestAccDynamoDBResourcePolicy_disappears
=== PAUSE TestAccDynamoDBResourcePolicy_disappears
=== CONT  TestAccDynamoDBResourcePolicy_basic
=== CONT  TestAccDynamoDBResourcePolicy_disappears
=== CONT  TestAccDynamoDBResourcePolicy_update
--- PASS: TestAccDynamoDBResourcePolicy_basic (39.27s)
--- PASS: TestAccDynamoDBResourcePolicy_update (64.17s)
--- PASS: TestAccDynamoDBResourcePolicy_disappears (64.71s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb	76.112s
```

**Pull request merged**